### PR TITLE
Add IR verification for use of non-value statement as values

### PR DESCRIPTION
**Pull request by Liozou**
As suggested by @Keno, this PR prevents non-value SSA statements from being used in value position. For instance, if `%3 = goto 5` is the third statement of the IR of a method, then any use of `%3` will result in an error.

The non-value statements are `GotoNode`, `GotoIfNot`, `ReturnNode`, `$(Expr(:leave, ...))` and `$(Expr(:gc_preserve_end, ...))`.

### CODE REVIEW

**Review by Keno — APPROVED** [review]

**Comment by kshyatt** [comment]
Should we re-run CI on this?

**Comment by ViralBShah** [comment]
Should be rebased on master, otherwise doctests will continue to fail on Travis. Rebasing should trigger the tests.

File: base/compiler/ssair/verify.jl [hunk 1]
```diff
@@ -71,6 +71,14 @@ function verify_ir(ir::IRCode)
     for (bb, idx) in bbidxiter(ir)
         stmt = ir.stmts[idx]
         stmt === nothing && continue
+        foreachssa(stmt) do ssa
+            x = ir.stmts[ssa.id]
+            if x isa GotoIfNot || x isa GotoNode || x isa ReturnNode ||
+               (x isa Expr && (x.head == :leave || x.head == :gc_preserve_end))
+                @verify_error "Non value statement $(ssa.id) (defined as $(ir.stmts[ssa.id])) used as a value in statement $idx: $stmt"
+                error()
```
Comment (vtjnash) [inline base/compiler/ssair/verify.jl]: @Keno, do you still have an issue open for fixing these post-alpha issues with the new optimizer (in this case, specifically referring to the use of `error()` being against code formatting guidelines for Base).

Also, when did the "Missing line number information" check below get disabled? If we aren't careful to track and set this on every line, the LLVM verifier will `abort` the program.

**Comment by fingolfin** [comment]
This was approved in 2018 by @Keno... do we still want it?

I've resolved the merge conflict, but otherwise know nothing about the IR verification code.

File: base/compiler/ssair/verify.jl [hunk 2]
```diff
@@ -254,6 +254,14 @@ function verify_ir(ir::IRCode, print::Bool=true,
         bb_unreachable(domtree, bb) && continue
         stmt = ir[SSAValue(idx)][:stmt]
         stmt === nothing && continue
+        foreachssa(stmt) do ssa
+            x = ir.stmts[ssa.id]
+            if x isa GotoIfNot || x isa GotoNode || x isa ReturnNode ||
+               (x isa Expr && (x.head == :leave || x.head == :gc_preserve_end))
+                @verify_error "Non value statement $(ssa.id) (defined as $(ir.stmts[ssa.id])) used as a value in statement $idx: $stmt"
+                error()
```
Comment (vchuravy) [inline base/compiler/ssair/verify.jl:262]:
```suggestion
                error("Verification error")
```

Comment (Liozou) [reply base/compiler/ssair/verify.jl:262]: I'm going to go for
```suggestion
                error("")
```
just to be consistent with the rest of the file

**Comment by Keno** [comment]
We already do this check: https://github.com/JuliaLang/julia/blob/master/base/compiler/ssair/verify.jl#L69

**Pull request closed without merging**